File: src/main/java/com/donrobo/fpbg/blueprint/building/AssemblingMachine2.java
package com.donrobo.fpbg.blueprint.building;

import net.sf.json.JSONObject;

public class AssemblingMachine2 extends AbstractBuilding {

    private final String recipe;

    public AssemblingMachine2(String recipe, int x, int y) {
        super(x, y);
        this.recipe = recipe;
    }

    @Override
    public String getName() {
        return "assembling-machine-2";
    }

    @Override
    public char getVisualizationCharacter() {
        return '2';
    }

    @Override
    public int getWidth() {
        return 3;
    }

    @Override
    public int getHeight() {
        return 3;
    }

    @Override
    public AssemblingMachine2 move(int x, int y) {
        return new AssemblingMachine2(getRecipe(), getX() + x, getY() + y);
    }

    public String getRecipe() {
        return recipe;
    }

    @Override
    protected double getBlueprintXOffset() {
        return 1;
    }

    @Override
    protected double getBlueprintYOffset() {
        return 1;
    }

    @Override
    protected void addCustomPropertiesToJson(JSONObject json) {
        json.put("recipe", recipe);
    }

    public static double getCraftingSpeed() {
        return 0.75;
    }
}

File: src/main/kotlin/com/donrobo/fpbg/blueprint/building/AssemblingMachine2.kt
package com.donrobo.fpbg.blueprint.building

import com.donrobo.fpbg.data.Int2
import net.sf.json.JSONObject

class AssemblingMachine2(val recipe: String, x: Int, y: Int) : AbstractBuilding(x, y) {

    override val name: String
        get() = "assembling-machine-2"

    override val visualizationCharacter: Char
        get() = '2'

    override val width: Int
        get() = 3

    override val height: Int
        get() = 3

    override fun move(x: Int, y: Int): AssemblingMachine2 {
        return AssemblingMachine2(recipe, this.x + x, this.y + y)
    }

    override val blueprintXOffset: Double
        get() = 1.0
    override val blueprintYOffset: Double
        get() = 1.0

    override fun addCustomPropertiesToJson(json: JSONObject) {
        json.put("recipe", recipe)
    }

    override fun rotateCCW(around: Int2, count: Int): Building {
        TODO("Untested")
//        val newPosition = position.rotateCCW(around, count) //probably some magic needed here
//
//        return AssemblingMachine2(recipe, newPosition.x, newPosition.y)
    }

    companion object {

        val craftingSpeed: Double
            get() = 0.75
    }
}
